Description: add "mediaType" to OCI index.json files
Author: Talon Bowler <talon.bowler@docker.com>
Origin: https://github.com/moby/buildkit/pull/4727 + https://github.com/containerd/containerd/pull/9867 (https://github.com/moby/buildkit/pull/4814)
Applied-Upstream: 0.14+

diff --git a/client/ociindex/ociindex.go b/client/ociindex/ociindex.go
index 512a77a68e67..5321f773d70e 100644
--- a/client/ociindex/ociindex.go
+++ b/client/ociindex/ociindex.go
@@ -102,6 +102,7 @@ func (s StoreIndex) Put(tag string, desc ocispecs.Descriptor) error {
 		}
 	}
 
+	setOCIIndexDefaults(&idx)
 	if err = insertDesc(&idx, desc, tag); err != nil {
 		return err
 	}
@@ -145,6 +146,19 @@ func (s StoreIndex) GetSingle() (*ocispecs.Descriptor, error) {
 	return nil, nil
 }
 
+// setOCIIndexDefaults updates zero values in index to their default values.
+func setOCIIndexDefaults(index *ocispecs.Index) {
+	if index == nil {
+		return
+	}
+	if index.SchemaVersion == 0 {
+		index.SchemaVersion = 2
+	}
+	if index.MediaType == "" {
+		index.MediaType = ocispecs.MediaTypeImageIndex
+	}
+}
+
 // insertDesc puts desc to index with tag.
 // Existing manifests with the same tag will be removed from the index.
 func insertDesc(index *ocispecs.Index, desc ocispecs.Descriptor, tag string) error {
@@ -152,9 +166,6 @@ func insertDesc(index *ocispecs.Index, desc ocispecs.Descriptor, tag string) err
 		return nil
 	}
 
-	if index.SchemaVersion == 0 {
-		index.SchemaVersion = 2
-	}
 	if tag != "" {
 		if desc.Annotations == nil {
 			desc.Annotations = make(map[string]string)
diff --git a/vendor/github.com/containerd/containerd/images/archive/exporter.go b/vendor/github.com/containerd/containerd/images/archive/exporter.go
index 1f17a3cdbfec..8513e9a8bf0f 100644
--- a/vendor/github.com/containerd/containerd/images/archive/exporter.go
+++ b/vendor/github.com/containerd/containerd/images/archive/exporter.go
@@ -471,6 +471,7 @@ func ociIndexRecord(manifests []ocispec.Descriptor) tarRecord {
 		Versioned: ocispecs.Versioned{
 			SchemaVersion: 2,
 		},
+		MediaType: ocispec.MediaTypeImageIndex,
 		Manifests: manifests,
 	}
 
